core/lib.rs

```diff
@@ -743,6 +748,45 @@ impl Connection {
 
         Ok(results)
     }
+
+    pub fn track_program(&self, program_rc: &Rc<Program>) {
+        let program_mut_ptr = Rc::as_ptr(program_rc) as *mut Program;
+        let new_node_ptr = NonNull::new(program_mut_ptr).unwrap();
+
+        let current_head = self.program_head.get();
+
+        unsafe {
+            new_node_ptr.as_ref().next_program.set(current_head);
+            new_node_ptr.as_ref().prev_program.set(None);
+
+            if let Some(old_head) = current_head {
+                old_head.as_ref().prev_program.set(Some(new_node_ptr));
+            }
+
+            self.program_head.set(Some(new_node_ptr));
+        }
+    }
```

Comment:
If we are following the route of linked list on this, can we have some tests for this? Regular linked list tests like we have in page_cache where we test unlinking head, tail, in the middle, fuzz tests etc...